src/ringcentral-web-phone.js

```diff
@@ -1155,6 +1171,237 @@
+    //------------------QOS---------------------------------//
+
+
+    /**
+     * @this {SIP.Session}
+     */
+    function publishQosStats(options){
+        var session = this;
+
+        if(session.qosStatsObj.status) {
+            session.getStatsResult && session.getStatsResult.nomore();
+            var networkType = calculateNetworkUsage(session) || '';
+            options = options || {};
+            var targetUrl = options.targetUrl || 'rtcpxr@rtcpxr.ringcentral.com:5060';
+            var event = options.event || 'vq-rtcpxr';
+            options.expires = 60;
+            options.contentType = "application/vq-rtcpxr";
+            options.extraHeaders = [];
+            options.extraHeaders.push('p-rc-client-info:' + 'cpuRC=0:0;cpuOS=0:0;netType=' + networkType + ';ram=0:0');
+
+            var body = createPublishBody(session);
+            var pub = session.ua.publish(targetUrl, event, body, options);
+            session.qosStatsObj.status = false;
+            pub.close();
+            session.emit('qos-published',body);
+        }
+        else{
+            session.logger.error('QOS collection not started');
+        }
+
+    }
+
+    /*--------------------------------------------------------------------------------------------------------------------*/
+
+    /**
+     * @this {SIP.Session}
+     */
+    function getQoSStatsTemplate() {
+        return {
+            localAddr: '',
+            remoteAddr: '',
+            callID: '',
+            localID: '',
+            remoteID: '',
+            origID: '',
+            fromTag:'',
+            toTag:'',
+            timestamp: {
+                start: '',
+                stop: ''
+            },
+
+            netType: '',
+
+            packetLost : 0,
+            packetsReceived: 0,
+
+            jitterBufferNominal: 0,
+            jitterBufferMax: 0,
+
+            jitterBufferDiscardRate: 0,
+
+            totalSumJitter:0,
+            totalIntervalCount:0,
+
+            NLR : 0,
+            JBM : 0 ,
+            JBN :  0,
+            JDR : 0,
+            MOSLQ: 0,
+
+            status : false
+        }
+    }
+
+    /*--------------------------------------------------------------------------------------------------------------------*/
+
+    /**
+     * @this {SIP.Session}
+     */
+    function startQosStatsCollection(){
+        var session =  this;
+        session.qosStatsObj.callID = session.request.call_id||'';
+        session.qosStatsObj.fromTag  = session.from_tag || '';
+        session.qosStatsObj.toTag  = session.to_tag || '';
+        session.qosStatsObj.localID = session.request.headers.From[0].raw||session.request.headers.From[0];
+        session.qosStatsObj.remoteID = session.request.headers.To[0].raw||session.request.headers.To[0];
+        session.qosStatsObj.origID = session.request.headers.From[0].raw||session.request.headers.From[0];
+        getStat(session);
```

Comment:
There's actually no need for second function `publishQosStats`, you can do all the stuff locally in this function.

```js
var qosStatsObj;
var netTypeObj;
var previousGetStatsResult;
getStats(peer, function (getStatsResult){
  // do things with qosStatsObj and netTypeObj
  previousGetStatsResult = getStatsResult;
});
session.on('terminate', function(){
  previousGetStatsResult.nomore();
  var calculatedStats =  calculateStats(qosStatsObj);
  var body = createPublishBody(calculatedStats);
});
```